### PR TITLE
125: Search invoice backend

### PR DESCRIPTION
**Pull request by ngocnguyenbk**
http://redmine.dtpebike.vn/issues/125

### CODE REVIEW

File: app/presenters/invoices_presenter.rb [hunk 1]
```diff
@@ -7,27 +7,45 @@ class InvoicesPresenter
...
-      total_count: total_count
+      total_count: total_count,
+      months: months
     }
   end
+
+  def months
+    return Rails.cache.read("months") if Rails.cache.read("months")
+
+    months = Invoice.all.pluck(:date_export).map(&:month).uniq.sort
```
Comment (ngocnguyenbk) [inline app/presenters/invoices_presenter.rb:47]: ???

File: app/javascript/component_invoices/searchForm.vue [hunk 2]
```diff
@@ -83,15 +93,49 @@ const {mapState, mapActions} = createNamespacedHelpers('invoice');
...
+    ...mapState(['params', 'resultMonths']),
+  },
+  mounted() {
+    this.roomNumber = this.params.room_number;
+    this.dateExport = this.params.date_export;
   },
   methods: {
     ...mapActions(['submitFormSearch']),
+    setMonth(month) {
+      for (const [_btn, element] of Object.entries(this.$refs)) {
```
Comment (ducmanh2111) [inline app/javascript/component_invoices/searchForm.vue:110]: Not need!

File: app/javascript/component_invoices/searchForm.vue [hunk 3]
```diff
@@ -83,15 +93,49 @@ const {mapState, mapActions} = createNamespacedHelpers('invoice');
...
+      } else {
+        button.blur();
+        this.dateExport = '';
+      }
+
+      this.submitForm();
+    },
+    submitForm() {
+      const page = 1;
+      const params = {};
```
Comment (ducmanh2111) [inline app/javascript/component_invoices/searchForm.vue:127]: params = {'room_number': ....}

File: app/javascript/mixins/convert_param_url.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import {forEach, replace} from 'lodash';
+
+export default {
+  methods: {
+    convert_params(params) {
+      let newParams = {};
+      forEach(params, function(value, key) {
+        const newKey = key.match(/^q\[(\w+)\]$/)[1];
```
Comment (ducmanh2111) [inline app/javascript/mixins/convert_param_url.js:8]: Please assign an regex a name as a constant

File: app/javascript/invoice.vue [hunk 5]
```diff
@@ -85,7 +88,11 @@ export default {
     },
   },
   created: function() {
-    this.$store.dispatch('invoice/getInvoices');
+    const searchParams = new URLSearchParams(location.search);
+    let params = Object.fromEntries(searchParams.entries());
+
+    params = this.convert_params(params);
```
Comment (ducmanh2111) [inline app/javascript/invoice.vue:94]: Please give params a new name not identical with old params

File: app/javascript/store/modules/invoice.js [hunk 6]
```diff
@@ -53,12 +54,20 @@ const state = {
 const actions = {
   submitFormSearch({commit}, payload) {
     commit('setParams', payload);
+    let params = { q: payload.params }
+    if(payload.page && payload.page != '1'){
+      params.page = payload.page
+    }
+
+    const path = '/invoices?' + jQuery.param(params)
```
Comment (ducmanh2111) [inline app/javascript/store/modules/invoice.js:62]: Interpolation using `${}`

File: app/javascript/component_invoices/searchForm.vue [hunk 7]
```diff
@@ -83,15 +93,37 @@ const {mapState, mapActions} = createNamespacedHelpers('invoice');
...
+    ...mapState(['params', 'resultMonths']),
+  },
+  mounted() {
+    this.roomNumber = this.params.room_number;
+    this.dateExport = this.params.date_export;
   },
   methods: {
     ...mapActions(['submitFormSearch']),
+    setMonth(month) {
+      this.dateExport = month;
```
Comment (ducmanh2111) [inline app/javascript/component_invoices/searchForm.vue:110]: Support clear dateExport when click to already focus button

File: app/javascript/component_invoices/searchForm.vue [hunk 8]
```diff
@@ -83,15 +93,44 @@ const {mapState, mapActions} = createNamespacedHelpers('invoice');
...
+  mounted() {
+    this.roomNumber = this.params.room_number;
+    this.dateExport = this.params.date_export;
   },
   methods: {
     ...mapActions(['submitFormSearch']),
+    setMonth(month) {
+      if (month == this.dateExport) {
+        this.dateExport = '';
+        const button = this.$refs[`button_${month}`][0];
```
Comment (ducmanh2111) [inline app/javascript/component_invoices/searchForm.vue:112]: remove line 112, 113

**Comment by ducmanh2111** [comment]
LGTM

**Pull request merged**